dhcpv6/option_bootfileparam.go

```diff
@@ -0,0 +1,68 @@
+package dhcpv6
+
+import (
+	"fmt"
+
+	"github.com/u-root/u-root/pkg/uio"
+)
+
+// OptBootFileParam implements the OptionBootfileParam option
+//
+// This module defines the OPT_BOOTFILE_PARAM structure.
+// https://www.ietf.org/rfc/rfc5970.txt (section 3.2)
+type OptBootFileParam []string
+
+var _ Option = OptBootFileParam(nil)
+
+// Code returns the option code
+func (op OptBootFileParam) Code() OptionCode {
+	return OptionBootfileParam
+}
+
+// ToBytes serializes the option and returns it as a sequence of bytes
+func (op OptBootFileParam) ToBytes() []byte {
+	buf := uio.NewBigEndianBuffer(nil)
+	for _, param := range op {
+		if len(param) >= 1<<16 {
+			panic("a too long parameter")
+		}
+		buf.Write16(uint16(len(param)))
+		buf.WriteBytes([]byte(param))
+		/*if err := buf.Error(); err != nil {
+			// TODO: description of `WriteBytes` says it could return
+			// an error via `buf.Error()`. But a quick look into implementation of
+			// `WriteBytes` at the moment of this comment showed it does not set any
+			// errors to `Error()` output. It's required to make a decision:
+			// to fix `WriteBytes`'s description or to find a way to handle an error here.
+		}*/
+	}
+	return buf.Data()
+}
+
+func (op OptBootFileParam) String() string {
+	return fmt.Sprintf("OptBootFileParam(%v)", ([]string)(op))
+}
+
+// ParseOptBootFileParam builds an OptBootFileParam structure from a sequence
+// of bytes. The input data does not include option code and length bytes.
+func ParseOptBootFileParam(data []byte) (result OptBootFileParam, err error) {
+	buf := uio.NewBigEndianBuffer(data)
+	for buf.Has(2) {
+		length := buf.Read16()
+		if err := buf.Error(); err != nil {
+			return nil, fmt.Errorf("[bootfile param] unable to read the parameter length: %w", err)
+		}
+
+		param, err := buf.ReadN(int(length))
+		if err != nil {
+			return nil, fmt.Errorf("[bootfile param] unable to read the parameter value: %w", err)
+		}
+
+		result = append(result, string(param))
+
+	}
+	if err := buf.FinError(); err != nil {
```

Comment:
Part of the idea of uio is that you only need to check this error here at the end, and any previously encountered error would be folded into this. So you don't need the previous 2 error checks, which should make the loop much simpler

Reply:
How it will exit the loop in case of error (if it's unable to consume the last byte)? I still have to add at least one error check into the loop. Also I would like to add explanation text to the error if possible (like `[bootfile param]`). However if you think I should as much drop error checks as possible anyway, then I will do it. Should drop one check?

Reply:
The only reason that Read16 or ReadN/CopyN would fail is because there aren't enough bytes left in the buffer.

buf.Has(2) should cover that. there should be no error check in the loop.

which check do you think still has to be there?

Reply:
@hugelgupf well, OK, but there's no description in the functions `Read*` that it's the only reason why it could fail, so there's no guarantee that it won't be changed in future. Anyway OK, I will fix it.